### PR TITLE
Fix sphinx warnings and restructure API docs

### PR DESCRIPTION
**Pull request by billbrod**
**Describe the change in this PR at a high-level**

This PR makes two related changes: restructure the API documentation and fix all warnings raised by sphinx while building the documentation.

The majority of the sphinx warnings were related to duplicate, missing, or mis-formatted anchors.

The API documentation restructure is inspired by [pynapple](https://github.com/pynapple-org/pynapple/blob/main/doc/api.rst) and [nemos](https://github.com/flatironinstitute/nemos/blob/main/docs/api_reference.rst) and swaps out the use of `sphinxcontrib-apidoc` to automatically generate the docs with a manual `api.rst`. This file explicitly lists all functions and classes that I expect users  want to use. It is paired with `api_modules.rst`, which generates the module pages linked from `api.rst`.

My goal with the API documentation is:
- As I said above, `api.rst` should include all functions and classes I expect users to use. They are grouped thematically and roughly ordered by utility, so that synthesis objects and their visualization helper functions are at the top, followed by models and metrics, then canonical computations, and finally miscellaneous tools.
- The synthesis objects each have their own dedicated documentation page. This page only contains the object with its attributes and methods in a table and then listed.
- The synthesis helper functions (e.g., `plot_synthesis_status`) live on a separate page. Each module (e.g., `metamer.py`) has a single page which lists all functions found in that module (in practice, everything except the synthesis objects).
- All other functions and objects have per-module pages. Most of them are ordered alphabetically, some by source.
    - I like having the per-module page (rather than per-function) because many of the functions/classes within a module are related. E.g., the models in `frontend.py` all build on each other, so it's useful to scroll up and down and see the other models.
- The synthesis objects' API docs list all attributes/methods (included inherited ones), whereas the models (which all inherit `torch.nn.Module`) only show the non-inherited members, since torch Modules have *so many* members.

**Link any related issues, discussions, PRs**

This addresses half of #360 (the points about things that should be linked). The rest will be addressed in a subsequent PR that switches the theme to pydata.

**Outstanding questions / particular feedback**

- What do you think of my API docs page structure? Does the order of the categories make sense?
- Right now, I am showing the signature for the synthesis objects and their helper functions, but not for anything else. I was doing that to show the consistency of the synthesis API, but I'm inclined to not show the signature for anything, because it's distracting. Thoughts?
- While grouping functions, it made me realize that the location of some functions doesn't make a lot of sense. E.g., `remove_grad` probably doesn't belong in `validate.py`, and `rectangular_to_polar` should probably be in `non_linearities.py`. I think I'm going to do that as part of #246, which will involve breaking the API anyway.
    - I think I will move functions/classes around so that the modules match the categories on this page, with the exception of moving `remove_grad`. So, do these categories (and their contents) make sense?
- Does doing a single docs page per synthesis object and then a docs page per module for everything else make sense? As I said above, I like having the per-module page (rather than per-function) because many of the functions/classes within a module are related. E.g., the models in `frontend.py` all build on each other, so it's useful to scroll up and down and see the other models. However, the synthesis objects themselves are important enough that I wanted them to stand alone.
- I removed `Synthesis` / `OptimizedSynthesis` from the documentation because they're abstract classes that I don't expect users to interact with. However, I do reference them in one of the docs page (`synthesis.md` / `Advanced Usage -> Synthesis object design`). Does this way of handling them make sense?
- There are some public objects that are not included in `api.rst`: the contents of `autodiff.py`. I should probably just make those all private, right? I don't expect users to call them.
  - Relatedly, should `Synthesis` / `OptimizedSynthesis` (both found in `synthesis.py`) be made private?
- Similarly, I'm not sure what to do with `fetch_data`. It is currently not exposed (users cannot access it at `plenoptic.data.fetch.fetch_data` but must instead explicitly import it `from plenoptic.data import fetch`). It is largely used internally, for grabbing files used in testing. However, there is one file that users may want: `portilla_simoncelli_images.tar.gz`, which includes the texture images from the Portilla-Simoncelli paper (used throughout our docs for that model). What should I do?

**Describe changes**

- Replaces auto-generated API docs (from `sphinxcontrib-apidoc`) with `docs/api.rst` (and `docs/api_modules.rst`).
    - `api.rst` is the user-facing page that is included in the documentation
    - `api_modules.rst` is an "orphan" page (not included in the toctree) and is just used to tell [autosummary](https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html) to generate the module pages we link to from `api.rst`
- Adds `-W -T` to sphinx makefile, so we always raise warnings as errors and display full error traceback.
- Turns on `nitpicky=True`, so that sphinx will complain about missing cross-references (and thus, paired with `-W`, raise an error)
- Adds new `linting/check_apidocs.py` to pre-commit (and thus our required checks). This ensures that all of plenoptic's public functions and classes are found in `docs/api.rst`
    - The inverse check (that everything in `api.rst` is part of plenoptic) is not done explicitly. Instead, it will result in a warning during the sphinx build, which (because of `-W`) will result in an error during the documentation build.
- In previous PRs when I had switched to using myst (#339 and #358), I had misunderstood the cross-reference syntax, so had a lot of e.g., ``{class}`Metamer <plenoptic.synthesize.metamer.Metamer>` ``. That is correct but overly verbose, so this PR replaces that with the more compact tilde syntax: ``{class}`~plenoptic.synthesize.metamer.Metamer` ``
- Updates `CONTRIBUTING.md` to reflect above: tell people they need to add new functions/modules to `api.rst` / `api_modules.rst` and updates example crossref syntax.
- in Makefile, remove the build directory and the auto-generated api docs (from [autosummary](https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html)) before the call to sphinx-build, to avoid cache issues
    - Moves myst-nb's jupyter cache folder outside the build directory so it doesn't get cleared as part of this
- Adds custom sphinx templates to handle synthesis objects, synthesis modules (without object), and modules where I want to preserve source order (rather than alphabetical). See top section for description of my goal here.
    - Similarly, tweaks the autodoc / autosummary options to get apidocs looking the way I like.
- Removes `numpydoc`, which is redundant with napoleon.
- Small restructuring of docs, adding explicit anchors.
- Removes `Synthesis` / `OptimizedSynthesis` from API docs entirely, since they are abstract classes. They are still referenced in `synthesis.md`, which describes the synthesis object APIs and how to implement your own.
- convergence functions now accept objects of type `Metamer | MetamerCTF | MADCompetition` instead of `OptimizedSynthesis` (since that is no longer a documented object)
- Changes `model_metric` to `model_metric_factory`, making it a loss factory (like `portilla_simoncelli_loss_factory` in #381), which is more appropriate.
- Lots of cross-reference warnings came from duplicate targets in docstrings of functions in the same module. E.g., both `ssim` and `ms_ssim` (in `perceptual_distance.py`) had references with the target `[1]_`. Changed that so that those targets are all unique and added citation to docstring, e.g., replacing `As described in [1]_` with `As described in Wang et al., 2004 [1]_`.
    - This is only a problem because I wanted to have the whole module on a single page in the docs. If I had separate pages for `ssim` and `ms_ssim`, they could've reused targets.
- Fixes small typos, errors, and mis-formatted links.
- Changes some functions from public to private, because users aren't expected to use them directly: `signal._interpolate1d`, `signal._raised_cosine`, `signal._steer` (all used by steerable pyramid).
- Moves `fisher_info_matrix_vector_product` and `fisher_info_matrix_eigenvalue` from `eigendistortion.py` to `autodiff.py`
- The entire `autodiff.py` module is not included in the API docs, it's for internal use only
- Documents `Metamer` and `MetamerCTF` `loss_function` attribute. It is not a property, because that would break saving behavior. But it now has a docstring and is included as a targetable anchor
- one confusing sphinx warning was because I had an extra pair of parentheses in a markdown link: `[title]((url))` instead of `[title](url)`. caused sphinx crossref to compain about a missing target.

**Checklist**

Affirm that you have done the following:

- [x] I have described the changes in this PR, following the template above.
- [x] I have added any necessary tests.
- [x] I have added any necessary documentation. This includes docstrings, updates to existing files found in `docs/`, or (for large changes) adding new files to the `docs/` folder.

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+   :heading-level: 3
+
+These functions perform some useful image-processing non-linearities on tensors or dictionaries of tensors.
+
+.. currentmodule:: plenoptic
+.. autosummary::
+   :signatures: none
+
+   ~tools.signal.rectangular_to_polar
+   ~tools.signal.polar_to_rectangular
```
[the remark this answers is not in the document]
Comment (billbrod) [reply docs/api.rst]: I think you're probably right here. I reworked this section a bit, adding a table summarizing the relationship between them.

File: docs/api.rst [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+   ~data.reptile_skin
+   ~data.color_wheel
+
+Plenoptic includes a helper function for downloading data. These files are used for our internal tests and documentation.
+
+.. autosummary::
+   :signatures: none
+
+   ~data.fetch.fetch_data
+   ~data.fetch.DOWNLOADABLE_FILES
```
[the remark this answers is not in the document]
Comment (billbrod) [reply docs/api.rst]: I have also changed imports so they can be accessed at `po.data.fetch_data`, and `po.data.DOWNLOADABLE_FILES`
